src/NuGet.Core/NuGet.Protocol/PackageDownloadContext.cs

```diff
@@ -37,12 +38,23 @@ public class PackageDownloadContext
             DirectDownloadDirectory = directDownloadDirectory;
         }
 
+        public PackageDownloadContext(
+            SourceCacheContext sourceCacheContext,
+            string directDownloadDirectory,
+            bool directDownload,
+            PackageNamespacesConfiguration packageNamespacesConfiguration)
+            : this(sourceCacheContext, directDownloadDirectory, directDownload)
+        {
+            PackageNamespacesConfiguration = packageNamespacesConfiguration;
```

Comment:
I strongly prefer to have only a single constructor that has a non-empty body, and all other constructors call it. In this case, I suggest adding the extra parameter to the existing constructor, and creating a new constructor with the old signature which invokes the first constructor, passing `null` as the value for the package namespace configuration.

Reply:
I refactored code little bit. Can you check now?

Reply:
I'm not sure if there was a misunderstanding or an oversight, but your change now has deleted a public API that customers and partners might be using. I don't think it makes sense to force everyone using NuGet.Protocol programatically to be forced to use package namespaces, so I think we should keep the old constructor signature.

I like that the first constructor in this class has an empty body, but we should add a third constructor, not break a public API.

Reply:
Ok. Addressed, I believe now it's correct implementation.

Reply:
my Q was directed to @zivkan

Reply:
> Is this a case of judgment call of which APIs we remove vs which APIs we keep?

Even though we're incrementing major version, which in SemVer allows breaking APIs, I think it would be much more considerate to customers to only delete APIs that were marked as obsolete for a few versions. Having said that, I didn't consider the fact that we're bumping major versions, but I don't think we should freely break APIs just because we are preparing for a major version.

In this scenario, when I [search github for PackageDownloadContext](https://github.com/search?q=PackageDownloadContext&type=code), I see lots of projects creating new instances of this class. And no, these search results are not all forks of NuGet.Client code. I don't know for certain that they're using this specific overload, but like I wrote above, unless it's been obsolete in a few shipped versions, I think it's kinder to add an overload. We can make the old overload obsolete if it makes sense, but consider it from a customer point of view. Don't mark it obsolete just because NuGet.Client isn't using it internally. Consider [these API design points](https://github.com/NuGet/NuGet.Client/blob/dev/docs/nuget-sdk.md#changes-to-apis) I wrote.

Again, in this scenario, [the .NET SDK does not use PackageDownloadContext](https://github.com/dotnet/sdk/search?q=PackageDownloadContext), but if we break any API that they're using, our insertions will fail and someone will either need to update their code to use our new API, or we'll be blocked on insertions until we make NuGet API compatible.


Finally, for this specific API, the PackageDownloadContext constructor, and whether or not the PackageNamespacesConfiguration should be mandatory, I vote no. I can easily imagine scenarios where customers (or even my own personal tools/evaluations) wants to programatically download packages, and not want to be forced to provide namespace configuration. Especially in scenarios where the program is downloading from a single source.